src/CoreFn/Expr.purs

```diff
@@ -131,69 +52,108 @@ data Expr a
+    "(Literal " <>
+      intercalate " " [ show a, show l ] <>
+    ")"
+  show (Constructor a t c fs) =
+    "(Constructor " <>
+      intercalate " " [ show a, show t, show c, show fs] <>
+    ")"
+  show (Accessor a s e) =
+    "(Accessor " <>
+      intercalate " " [ show a, show s, show e ] <>
+    ")"
+  show (ObjectUpdate a e fs) =
+    "(ObjectUpdate " <>
+      intercalate " " [ show a, show e, show fs ] <>
+    ")"
+  show (Abs a i e) =
+    "(Abs " <>
+      intercalate " " [ show a, show i, show e ] <>
+    ")"
+  show (App a e1 e2) =
+    "(App " <>
+      intercalate " " [ show a, show e1, show e2 ] <>
+    ")"
+  show (Var a q) =
+    "(Var " <>
+      intercalate " " [ show a, show q ] <>
+    ")"
+  show (Case a es cs) =
+    "(Case " <>
+      intercalate " " [ show a, show es, show cs ] <>
+    ")"
+  show (Let a bs e) =
+    "(Let " <>
+      intercalate " " [ show a, show bs, show e ] <>
+    ")"
+
 
 -- |
 --  A let or module binding.
 --
-data Bind a = Bind (Array (Tuple (Tuple a Ident) (Expr a)))
+data Bind a
+  = NonRec (Bind' a)
+  | Rec (Array (Bind' a))
+
+type Bind' a = Tuple (Tuple a Ident) (Expr a)
 
 derive instance eqBind :: Eq a => Eq (Bind a)
+-- derive instance functorBind :: Functor Bind
 derive instance ordBind :: Ord a => Ord (Bind a)
 
 instance showBind :: Show a => Show (Bind a) where
-  show (Bind x) = "(Bind " <> show x <> ")"
-
-readBind :: Foreign -> F (Bind Unit)
-readBind x = do
-  pairs <- objectProps x
-  bindings <- traverse fromPair pairs
-  pure $ Bind bindings
-
-  where
-
-  fromPair
-    :: { key :: String, value :: Foreign }
-    -> F (Tuple (Tuple Unit Ident) (Expr Unit))
-  fromPair pair = do
-    expr <- readExpr pair.value
-    let ident = Ident pair.key
-    pure $ Tuple (Tuple unit ident) expr
-
-readBindJSON :: String -> F (Bind Unit)
-readBindJSON = parseJSON >=> readBind
+  show (NonRec b) = "(NonRec " <> show b <> ")"
+  show (Rec b) = "(Rec " <> show b <> ")"
+
+
+-- |
+-- A guard is just a boolean-valued expression that appears alongside a set of binders
+--
+type Guard a = Expr a
+
+
+-- |
+-- An alternative in a case statement
+--
+data CaseAlternative a = CaseAlternative
+  {
+    -- |
+    -- A collection of binders with which to match the inputs
+    caseAlternativeBinders :: (Array (Binder a))
+    -- |
+    -- The result expression or a collect of guarded expressions
+  , caseAlternativeResult :: (Either (Array (Tuple (Guard a) (Expr a))) (Expr a))
+  }
+
+derive instance eqCaseAlternative :: Eq a => Eq (CaseAlternative a)
+derive instance ordCaseAlternative :: Ord a => Ord (CaseAlternative a)
+
+-- instance functorCaseAlternative :: Functor CaseAlternative where
+--   map f (CaseAlternative { caseAlternativeBinders, caseAlternativeResult }) = CaseAlternative
+--     { caseAlternativeBinders: (map (map f) caseAlternativeBinders)
+--     , caseAlternativeResult: (either (Left <<< map (map f *** map f)) (Right <<< map f) caseAlternativeResult)
+--     }
```

Comment:
I think the last thing to do in this PR is to figure out why the Functor instances that are commented out don't work.

They're not actually used anywhere but it might be nice to have them.

This was taken from https://github.com/purescript/purescript/blob/c19a79285c5697b59c46803483b41ffeaa5cd7b0/src/Language/PureScript/CoreFn/Expr.hs#L89-L93

Reply:
I don't see a Functor instance for `Expr`.

Reply:
The other instances are here:

https://github.com/paulyoung/purescript-corefn/blob/aa1d7362292cd6f0b224631ee802a1081bbe23dc/src/CoreFn/Expr.purs#L65

https://github.com/paulyoung/purescript-corefn/blob/aa1d7362292cd6f0b224631ee802a1081bbe23dc/src/CoreFn/Expr.purs#L117

Reply:
This is the error:

```
[1/1 TypesDoNotUnify] src/CoreFn/Expr.purs:117:8
  117  derive instance functorBind :: Functor Bind
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  Could not match type
  
    b1
  
  with type
  
    a2
  
  while trying to match type Expr b1
    with type Expr t0
  while checking that expression (map (map $f146)) $v148
    has type Tuple (Tuple t0 Ident) (Expr t0)
  in value declaration functorBind
  
  where b1 is a rigid type variable
        a2 is a rigid type variable
        t0 is an unknown type
```

https://travis-ci.org/paulyoung/purescript-corefn/jobs/345111990#L1549

I guess the issue might just be that `Functor` can’t be derived, all though I’m not sure if there would be a more specific error if that was the case.